### PR TITLE
build: change vllm version to github.com/opendatahub-io/vllm

### PR DESCRIPTION
**Pull request by alimaredia**
This PR switches to using a hardened branch of vllm that makes it easier to align dependencies on, and has been tested to work well in combination with future the SDG improvements. In the future we may provide a version of these requirements that tracks with the latest upstream

**Checklist:**

- [x] **Commit Message Formatting**: Commit titles and messages follow guidelines in the
  [conventional commits](https://www.conventionalcommits.org/en/v1.0.0/#summary).
- [ ] [Changelog](https://github.com/instructlab/instructlab/blob/main/CHANGELOG.md) updated with breaking and/or notable changes for the next minor release.
- [ ] Documentation has been updated, if necessary.
- [ ] Unit tests have been added, if necessary.
- [ ] Integration tests have been added, if necessary.

### CODE REVIEW

**Review by tiran — CHANGES_REQUESTED** [review]
We cannot upload a package to PyPI that has a requirement with an URL. You have to drop the optional dependency from pyproject.toml.

**Comment by n1hility** [comment]
@tiran ah yes good catch. Dropping as an official optional dep makes sense to me for now.

**Comment by alimaredia** [comment]
@n1hility Since dropping the optional dependency gets rid of the installation option of `pip install instructlab[vllm]` for a users vllm workflow, should we document that users need to install the vllm specified here separately?

**Comment by nathan-weinberg** [comment]
I was talking with @JamesKunstle and it seems upstream vLLM is only using `scipy` in one place: https://github.com/vllm-project/vllm/blob/c5832d2ae9431a1672d547c232ec46b1a9051ff0/vllm/attention/ops/blocksparse_attention/utils.py#L22

I think ideally we could pull that requirement (assuming there is an equivalent way to do that single op) and continue to use upstream vLLM instead of a fork. We made a similar ask for the `lm_eval` lib and got warm reception: https://github.com/EleutherAI/lm-evaluation-harness/issues/2059

**Comment by n1hility** [comment]
> I was talking with @JamesKunstle and it seems upstream vLLM is only using `scipy` in one place: https://github.com/vllm-project/vllm/blob/c5832d2ae9431a1672d547c232ec46b1a9051ff0/vllm/attention/ops/blocksparse_attention/utils.py#L22
> 
> I think ideally we could pull that requirement (assuming there is an equivalent way to do that single op) and continue to use upstream vLLM instead of a fork. We made a similar ask for the `lm_eval` lib and got warm reception: [EleutherAI/lm-evaluation-harness#2059](https://github.com/EleutherAI/lm-evaluation-harness/issues/2059)

There's a few other differences in that branch that we want. Although, the intention of it is to track closely to upstream. So it's not a full fork, but rather syncing one. I think we'd want upstream support at some point, but for now we are trying to reduce the number of permutations we have while we complete the remaining pieces.

**Review by n1hility — APPROVED** [review]

**Review by nathan-weinberg — APPROVED** [review]

**Comment by n1hility** [comment]
> @n1hility Since dropping the optional dependency gets rid of the installation option of `pip install instructlab[vllm]` for a users vllm workflow, should we document that users need to install the vllm specified here separately?

Yeah I wonder if what we need to do is document a command that looks like:

```
pip install git+https://github.com/instructlab/instructlab@version -r github.com/instructlab/instructlab/tree/version requirements.txt -r github.com/instructlab/instructlab/tree/version requirements-vllm-cuda.txt
```

**Comment by alimaredia** [comment]
Putting this on hold until I can verify that someone can test by hand ilab working with this version of vLLM.

**Comment by nathan-weinberg** [comment]
I'd like if we can have a followup issue (maybe in https://github.com/opendatahub-io/vllm would be appropriate place for this) to get that midstream vLLM implementation to do proper PyPI packaging and publishing so we're not installing a `git`-based checkpoint - what do you think @n1hility?

**Review by russellb — CHANGES_REQUESTED** [review]
I would prefer we get #1559 merged and functional first before any further big vllm changes. We need testing in place that can do some basic validation.

**Review by leseb — COMMENTED** [review]
Please update your commit message with the PR description. Thanks!

**Comment by danmcp** [comment]
> Putting this on hold until I can verify that someone can test by hand ilab working with this version of vLLM.

I have verified this PR works with the changes here: https://github.com/instructlab/instructlab/pull/1589

**Comment by n1hility** [comment]
@Mergifyio rebase

**Comment by n1hility** [comment]
@Mergifyio rebase

**Review by leseb — COMMENTED** [review]
Please update your commit message referencing the content of the PR description. Thanks

**Comment by alimaredia** [comment]
@leseb updated the commit message.

**Review by leseb — APPROVED** [review]

**Review by danmcp — APPROVED** [review]

**Review by russellb — APPROVED** [review]

File: requirements-vllm-cuda.txt [hunk 1]
```diff
@@ -2,9 +2,4 @@
 # Dependencies for installing vLLM on CUDA
 
 # vLLM only supports Linux platform (including WSL)
-vllm>=0.5.0; sys_platform == 'linux' and platform_machine == 'x86_64' and python_version >= '3.9' and python_version <= '3.11'
-
-# Most the dependencies below were taken from https://github.com/vllm-project/vllm/blob/main/requirements-cuda.txt
```
Comment (leseb) [inline requirements-vllm-cuda.txt:7]: I believe these are still needed, why did they go away?

**Review by leseb — COMMENTED** [review]
Added the hold label since I have a small question, but this feels ready!

File: requirements-vllm-cuda.txt [hunk 2]
```diff
@@ -2,9 +2,4 @@
...
-vllm-flash-attn == 2.5.9 ; sys_platform == 'linux' and platform_machine == 'x86_64' and python_version >= '3.9' and python_version <= '3.11'
-nvidia-ml-py3; sys_platform == 'linux' and platform_machine == 'x86_64' and python_version >= '3.9' and python_version <= '3.11'
-xformers == 0.0.26.post1; sys_platform == 'linux' and platform_machine == 'x86_64' and python_version >= '3.9' and python_version <= '3.11'
+vllm @git+https://github.com/opendatahub-io/vllm@2024.06.25 ; sys_platform == 'linux' and platform_machine == 'x86_64' and python_version >= '3.9' and python_version <= '3.11'
```
Comment (leseb) [inline requirements-vllm-cuda.txt:5]: from @tiran offline

> You can drop the python version range. Our package metadata already limits to 3.10 and 3.11

**Review by leseb — CHANGES_REQUESTED** [review]

File: requirements-vllm-cuda.txt (hunk 1 above)
Comment (tiran) [reply requirements-vllm-cuda.txt:7]: vLLM's `setup.py` picks up the dependencies automatically: https://github.com/opendatahub-io/vllm/blob/main/setup.py#L367

**Review by leseb — COMMENTED** [review]
Removing my request change since the request did not come from me.

File: requirements-vllm-cuda.txt (hunk 2 above)
Comment (tiran) [inline requirements-vllm-cuda.txt:5]:
```suggestion
vllm @git+https://github.com/opendatahub-io/vllm@2024.06.25 ; sys_platform == 'linux' and platform_machine == 'x86_64'
```

**Comment by tiran** [comment]
> Yeah I wonder if what we need to do is document a command that looks like:
> 
> ```
> pip install git+https://github.com/instructlab/instructlab@version \ 
>  -r https:///raw.githubusercontent.com/instructlab/instructlab/version/requirements.txt  \
>  -r https:///raw.githubusercontent.com/instructlab/instructlab/version/requirements-vllm-cuda.txt
> ```

To install InstructLab with vLLM on CUDA:
```
VERSION=0.18.0
pip install instructlab==${VERSION} -r https://raw.githubusercontent.com/instructlab/instructlab/v${VERSION}/requirements.txt
```

You don't need `requirements.txt`, because that information is baked into the wheel. The path to the vLLM CUDA txt file needs `v0.18.0` (with a leading `v`).

**Comment by alimaredia** [comment]
@leseb when you feel as if the PR is ready, I think the only thing preventing it from merging is the hold label you added.

**Pull request merged**